File: src/modlunky2/sprites/deco_sheet.py
from .base_classes.base_deco_sheet import AbstractDecoSheet

__all__ = [
    "CaveDecoSheet",
    "VolcanaDecoSheet",
    "JungleDecoSheet",
    "TidePoolDecoSheet",
    "TempleDecoSheet",
    "IceCavesDecoSheet",
    "BabylonDecoSheet",
    "SunkenCityDecoSheet",
    "EggplantDecoSheet",
    "SurfaceDecoSheet",
]


class CaveDecoSheet(AbstractDecoSheet):
    biome_name = "cave"
    _chunk_size = 128
    _chunk_map = {"kali_bg": (0, 0, 4, 5), "woodenlog_trap": (1, 5, 3, 10)}


class VolcanaDecoSheet(AbstractDecoSheet):
    biome_name = "volcano"
    _chunk_size = 128
    _chunk_map = {
        "kali_bg": (0, 0, 4, 5),
        "drill": (1, 5, 3, 8),
        "vlad_banner": (1, 8, 3, 12),
    }


class JungleDecoSheet(AbstractDecoSheet):
    biome_name = "jungle"
    _chunk_size = 128
    _chunk_map = {"stone_door": (0, 5, 3, 7)}


class TidePoolDecoSheet(AbstractDecoSheet):
    biome_name = "tidepool"
    _chunk_size = 128
    _chunk_map = {}


class TempleDecoSheet(AbstractDecoSheet):
    biome_name = "temple"
    _chunk_size = 128
    _chunk_map = {}


class IceCavesDecoSheet(AbstractDecoSheet):
    biome_name = "ice"
    _chunk_size = 128
    _chunk_map = {
        "moai_statue": (0, 7, 4, 12),
    }


class BabylonDecoSheet(AbstractDecoSheet):
    biome_name = "babylon"
    _chunk_size = 128
    _chunk_map = {}


class SunkenCityDecoSheet(AbstractDecoSheet):
    biome_name = "sunken"
    _chunk_size = 128
    _chunk_map = {
        "mother_statue": (0, 5, 2, 11),
    }


class EggplantDecoSheet(AbstractDecoSheet):
    biome_name = "eggplant"
    _chunk_size = 128
    _chunk_map = {}


class SurfaceDecoSheet(AbstractDecoSheet):
    biome_name = "basecamp"
    _chunk_size = 128
    _chunk_map = {
        "construction_sign": (5, 12, 6, 13),
        "boombox": (5, 14, 6, 15),
        "tutorial_speedrun_sign": (7, 14, 8, 15),
        "tutorial_menu_sign": (7, 14, 8, 15),
        "singlebed": (0, 15, 1, 16),
        "dresser": (2, 14, 3, 15),
        "bunkbed": (0, 13, 2, 15),
        "diningtable": (14, 10, 16, 11),
        "sidetable": (9, 12, 10, 13),
        "chair_looking_left": (3, 15, 4, 16),
        "chair_looking_right": (3, 15, 4, 16),
        "couch": (6, 15, 8, 16),
        "tv": (8, 15, 9, 16),
        "dog_sign": (12, 10, 14, 12),
        "shortcut_station_banner": (10, 8, 12, 12),
        "telescope": (12, 8, 14, 10),
    }

File: crates/ml2_sprites/src/data/loaders/deco_sheet.rs
// Sprite loader/merger configs describing which chunks live in which
// sheet, keyed by tile name for the sprite fetcher lookup.

use crate::LoaderConfig;
use crate::data::chunks;

pub fn cave_deco_sheet() -> LoaderConfig {
    LoaderConfig {
        name: "CaveDecoSheet".into(),
        sprite_sheet_path: "Data/Textures/deco_cave.png".into(),
        chunk_size: 128,
        chunk_map: chunks(&[
            ("kali_bg", 0.0, 0.0, 4.0, 5.0),
            ("woodenlog_trap", 1.0, 5.0, 3.0, 10.0),
            ("udjat_wall_heads", 8.0, 8.0, 12.0, 12.0),
        ]),
    }
}

pub fn volcana_deco_sheet() -> LoaderConfig {
    LoaderConfig {
        name: "VolcanaDecoSheet".into(),
        sprite_sheet_path: "Data/Textures/deco_volcano.png".into(),
        chunk_size: 128,
        chunk_map: chunks(&[
            ("kali_bg", 0.0, 0.0, 4.0, 5.0),
            ("drill", 1.0, 5.0, 3.0, 8.0),
            ("vlad_banner", 1.0, 8.0, 3.0, 12.0),
        ]),
    }
}

pub fn jungle_deco_sheet() -> LoaderConfig {
    LoaderConfig {
        name: "JungleDecoSheet".into(),
        sprite_sheet_path: "Data/Textures/deco_jungle.png".into(),
        chunk_size: 128,
        chunk_map: chunks(&[("stone_door", 0.0, 5.0, 3.0, 7.0)]),
    }
}

pub fn ice_caves_deco_sheet() -> LoaderConfig {
    LoaderConfig {
        name: "IceCavesDecoSheet".into(),
        sprite_sheet_path: "Data/Textures/deco_ice.png".into(),
        chunk_size: 128,
        chunk_map: chunks(&[
            ("moai_statue", 0.0, 7.0, 4.0, 12.0),
            ("boulder", 0.0, 5.0, 2.0, 7.0),
        ]),
    }
}

pub fn sunken_city_deco_sheet() -> LoaderConfig {
    LoaderConfig {
        name: "SunkenCityDecoSheet".into(),
        sprite_sheet_path: "Data/Textures/deco_sunken.png".into(),
        chunk_size: 128,
        chunk_map: chunks(&[("mother_statue", 0.0, 5.0, 2.0, 11.0)]),
    }
}

pub fn surface_deco_sheet() -> LoaderConfig {
    LoaderConfig {
        name: "SurfaceDecoSheet".into(),
        sprite_sheet_path: "Data/Textures/deco_basecamp.png".into(),
        chunk_size: 128,
        chunk_map: chunks(&[
            ("construction_sign", 5.0, 12.0, 6.0, 13.0),
            ("boombox", 5.0, 14.0, 6.0, 15.0),
            ("tutorial_speedrun_sign", 7.0, 14.0, 8.0, 15.0),
            ("tutorial_menu_sign", 7.0, 14.0, 8.0, 15.0),
            ("singlebed", 0.0, 15.0, 1.0, 16.0),
            ("dresser", 2.0, 14.0, 3.0, 15.0),
            ("bunkbed", 0.0, 13.0, 2.0, 15.0),
            ("diningtable", 14.0, 10.0, 16.0, 11.0),
            ("sidetable", 9.0, 12.0, 10.0, 13.0),
            ("chair_looking_left", 3.0, 15.0, 4.0, 16.0),
            ("chair_looking_right", 3.0, 15.0, 4.0, 16.0),
            ("couch", 6.0, 15.0, 8.0, 16.0),
            ("tv", 8.0, 15.0, 9.0, 16.0),
            ("dog_sign", 12.0, 10.0, 14.0, 12.0),
            ("shortcut_station_banner", 10.0, 8.0, 12.0, 12.0),
            ("telescope", 12.0, 8.0, 14.0, 10.0),
        ]),
    }
}
